From: Eric Dumazet <edumazet@google.com>
Date: Fri, 15 Sep 2023 17:11:11 +0000
Subject: scsi: iscsi_tcp: restrict to TCP sockets
Git-commit: f4f82c52a0ead5ab363d207d06f81b967d09ffb8
Patch-mainline: v6.6-rc3
References: git-fixes

Nothing prevents iscsi_sw_tcp_conn_bind() to receive file descriptor
pointing to non TCP socket (af_unix for example).

Return -EINVAL if this is attempted, instead of crashing the kernel.

[lduncan: hard-coded sk_is_tcp() since it doesn't exist here]

Fixes: 7ba247138907 ("[SCSI] open-iscsi/linux-iscsi-5 Initiator: Initiator code")
Signed-off-by: Eric Dumazet <edumazet@google.com>
Cc: Lee Duncan <lduncan@suse.com>
Cc: Chris Leech <cleech@redhat.com>
Cc: Mike Christie <michael.christie@oracle.com>
Cc: "James E.J. Bottomley" <jejb@linux.ibm.com>
Cc: "Martin K. Petersen" <martin.petersen@oracle.com>
Cc: open-iscsi@googlegroups.com
Cc: linux-scsi@vger.kernel.org
Reviewed-by: Mike Christie <michael.christie@oracle.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Lee Duncan <lduncan@suse.com>
---
 drivers/scsi/iscsi_tcp.c |    6 ++++++
 1 file changed, 6 insertions(+)

--- a/drivers/scsi/iscsi_tcp.c
+++ b/drivers/scsi/iscsi_tcp.c
@@ -679,6 +679,12 @@ iscsi_sw_tcp_conn_bind(struct iscsi_cls_
 		return -EEXIST;
 	}
 
+	err = -EINVAL;
+	/* hardcode sk_is_tcp() since it is not available */
+	if (!(sock->sk->sk_type == SOCK_STREAM &&
+	      sock->sk->sk_protocol == IPPROTO_TCP))
+		goto free_socket;
+
 	err = iscsi_conn_bind(cls_session, cls_conn, is_leading);
 	if (err)
 		goto free_socket;
